013-generalize-address-spaces.md

```diff
@@ -2,429 +2,372 @@
+first paragraph:
+
+>    `DW_ASPACE_default` is defined as the target architecture default
+>    address space.  See 2.12 Address Spaces.
 
 After the definition of `DW_OP_addrx` add:
 
-> 3.  `DW_OP_form_aspace_address`  
->     `DW_OP_form_aspace_address` pops top two stack entries. The first must be
->     an integral type value that represents a target architecture specific
->     address space identifier AS. The second must be an integral type value
->     that represents an address A.
-> 
->     The address size S is defined as the address bit size of the target
->     architecture specific address space that corresponds to AS.
-> 
->     A is adjusted to S bits by zero extending if necessary, and then
->     treating the least significant S bits as an unsigned value A'.
-> 
->     It pushes a location description L with one memory location description
->     SL on the stack. SL specifies the memory location storage LS that
->     corresponds to AS with a bit offset equal to A' scaled by 8 (the byte
->     size).
-> 
->     If AS is an address space that is specific to context elements, then LS
->     corresponds to the location storage associated with the current context.
-> 
->     [non-normative] For example, if AS is for per thread storage then LS is
->     the location storage for the current thread. Therefore, if L is accessed
->     by an operation, the location storage selected when the location
->     description was created is accessed, and not the location storage
->     associated with the current context of the access operation.
-> 
->     The DWARF expression is ill-formed if AS is not one of the values
->     defined by the target architecture specific `DW_ASPACE_*` values.
-> 
->     See Section 2.5.2.3 "Implicit Locations" for
->     special rules concerning implicit pointer values produced by
->     dereferencing implicit location descriptions created by the
->     `DW_OP_implicit_pointer` and `DW_OP_aspace_implicit_pointer` operations.
+>    3. `DW_OP_mem` 
+>       <[integral] A> <[integral] AS> → <[memory location] L>
+>
+>    `DW_OP_mem` pops top two stack entries. The first must be an
+>    integral type value that represents a target architecture
+>    specific address space identifier AS. The second is also an
+>    integral type which represents the address A the offset into that
+>    address space.
+>
+>    The address size S is defined as the address bit size of the
+>    target architecture specific address space that corresponds to
+>    AS.
+>
+>    A is adjusted to S bits by zero extending if necessary, and then
+>    treating the least significant S bits as an unsigned value A'.
+>
+>    It pushes a memory location within the address space AS whose
+>    offset is A.
+>
+>    If AS is an address space that is specific to context elements,
+>    then L corresponds to the location storage associated with the
+>    current context when the location is created, not the context
```

Comment:
"when the location is created, not the context when that location is used." -> "when the DW_OP_mem operation is evaluated, not the context when that location returned by the evaluation of DW_OP_mem is used."